### ISSUE TITLE
Page Model not yet loaded in custom route

### ISSUE THREAD

**Issue by ovenum** [k-next/kirby#4591]
## Description
When registering a custom route with the callback pattern Kirby does not apply the correct page model when using the `page` helper with a preceding call to `childrenAndDrafts()`.

```php
// Registering routes inside a plugin
Kirby\Cms\App::plugin('plugindomain/pluginname', [
   'routes' => function (Kirby\Cms\App $kirby) {
       // Removing this line will apply the correct page model when calling page() from within in the route callback
        $pageWithChildren = $kirby->site()->find('pageWithChildren')->childrenAndDrafts();
    
        return [
            [
                'pattern' => '(:any)',
                'action' => function ($slug) {
                    
                    // This will return a Kirby\Cms\Page object and not the expected model
                    $page = page('pageWithChildren/' . $slug);

                    if ($page) {
                        return kirby()->site()->visit($page);
                    }

                    $this->next();
                }
            ]
        ];
    },
]);
```

**Expected behavior**  
The correct page model should be applied inside the route handler when using the `page` helper function.

## To reproduce
See code above and create template, blueprint and page model for the child pages.

## Your setup
**Kirby Version**  
3.7.2.1

**Additional context**
`Kirby\Cms\Page::childrenAndDrafts()` or `Kirby\Cms\Page::children()` will call `Kirby\Cms\Page::inventory()` setting the page `inventory` property on its first call, but not loading the child pages models.

```php
$pageWithChildren->inventory()['children'];
// Will return something like this
[
    [
        'dirname' => 'childpageDir',
        'model' => null,  // Childpage model not loaded
        'num' => 'childpageNum',
        'root' => 'childpageRoot',
        'slug' => 'childpageSlug'
    ],
    …
]

```

**Comment by ovenum** [comment]
After further stepping through the code i found the reason. It’s too early to call `Kirby\Cms\Page::inventory()`. 
`Kirby\Cms\Page::$models` is still an empty array when called from within the route callback function inside a plugin. 

Could it be more beneficial to register page models before the routes?

**Comment by lukasbestle** [comment]
To be honest I think this is going to be hard to fix, if it's even possible.

Kirby first loads all plugin extensions (`extensionsFromPlugins`) and then all site extensions (`extensionsFromFolders`):

https://github.com/getkirby/kirby/blob/44158e8a5856384f35b07263240abc96db16ae7e/src/Cms/App.php#L135-L140

This means that e.g. if you register routes from a plugin but page models from the site, Kirby will first call the routes closure before the page models from the site are even considered.

Kirby needs this order so that site extensions always override plugin extensions. This priority order is essential so that plugins can be customized by the site developer.

**Comment by ovenum** [comment]
Thanks for looking into this @lukasbestle.

> This means that e.g. if you register routes from a plugin but page models from the site, Kirby will first call the routes closure before the page models from the site are even considered.

That is exactly what happened.

Maybe Kirby could notify/ throw an Exception when `Page::Models` or any function depending on it is used before complete initialisation? Once `Kirby\Cms\Page::inventory()` was called, `$inventory` is set and cached for the rest of the Request. 

https://github.com/getkirby/kirby/blob/44158e8a5856384f35b07263240abc96db16ae7e/src/Cms/Page.php#L524-L529

This will break all page models within $inventory that are registered after the first call to inventory(). Kirby just does not know them yet and will fall back to the default Kirby\Cms\Page as model.

This can break a complete page when, for example, `site()->find()` is called too early.

**Comment by lukasbestle** [comment]
I agree we need a way to track the loading state so that parts that depend on other parts being loaded can assert the current state.

We could introduce a new `int $app->loadingState` prop with an internal `$app->assertLoadingState(int $state, string $message)` method. This method would check if the state is at least what's passed, otherwise there would be an exception.

The state would be an int that starts at 0 and gets incremented as more and more parts of the system are loaded. Alternatively we could make it an array that gets values for the different parts that are loaded. That would be more flexible, but maybe not needed.

**Comment by distantnative** [comment]
It would also prevent current use cases that might depend on already calling $site->index() even if not all models have been loaded as they don't depend on models. Don't think we'd be really solving the problem, just pushing it elsewhere :/

**Comment by lukasbestle** [comment]
But we cannot allow early access for the site object and its children if we cache the objects in memory. Then we need to handle it internally so that caches are cleared and objects are consistent with the loading state at any time. But that would reduce performance for all while only benefitting few users.

**Comment by ovenum** [comment]
> It would also prevent current use cases that might depend on already calling $site->index() even if not all models have been loaded as they don't depend on models. Don't think we'd be really solving the problem, just pushing it elsewhere :/

If site()->index() throws an Exception, wouldn’t a try/catch block wrapped around it still  ›_allow_‹ these cases with the added bonus that it’s being done with the knowledge of its limitations when called too early?

**Comment by distantnative** [comment]
I'm not saying that the current state is good, but I'm worried that breaking existing use cases that way for solving maybe rather edge cases might bite us even more.

**Comment by distantnative** [comment]
I'd also think that we have the issue between different plugins where one is loaded before the other which only defines a certain model.

**Comment by lukasbestle** [comment]
We had so many of these "order of loading" issues over the years and so far we have always tried to solve them by changing the loading order. But of course this breaks other parts. So I think we really need to acknowledge the internal dependencies and communicate them clearly. Every call to a part of Kirby that is not loaded should result in an error. An error is much much better than broken behavior that can only be found after a long debugging session.

**Comment by distantnative** [comment]
Maybe, but I am afraid this ends in a principle debate without any actual results. We need to consider not just what's ideal but what's feasible and realistic to be done anytime soon.

**Comment by lukasbestle** [comment]
Such internal loading state assertions are definitely a feature we could implement. It takes time like with any other feature, but I wouldn't say it's infeasible.

**Comment by distantnative** [comment]
When I say feasible, I don't mean impossible - just given our resources and priorities we have quite a few of these topics that keep being discussed here in the issues but then never happen because of limited resources and priorities. I'd like to be realistic about expectations - and I don't see this happening anytime soon, at least from my side.

**Comment by lukasbestle** [comment]
Certainly. I still think it makes sense to discuss such topics and develop ideas over time. I think we should rather discuss if this is a direction we want to go in generally.
